### PR TITLE
Update DefExpression after value propagation

### PR DESCRIPTION
**Pull request by ptomin**
Update `DefExpression` after value propagation

### CODE REVIEW

File: subjects/regressions/reko-90/PP.c [hunk 1]
```diff
@@ -3233,29 +3225,23 @@ void fn0800_2688(word16 bp, selector ds, word16 wArg02, word16 wArg04, word16 wA
 l0800_2729:
 l0800_272E:
 	int32 dx_ax_177 = (int32) SLICE(SEQ(ax_142, ax_134 + wArg02) + _, word16, 16)
-	word16 bx_185 = (word16) ptrLoc06
-	selector es_186 = SLICE(ptrLoc06, selector, 16)
 	word16 dx_181
-	Mem187[es_186:bx_185 + 0x00:byte] = (byte) fn0800_8C8A((word16) dx_ax_177, 0x08, SLICE(dx_ax_177, word16, 16), bp, out dx_181)
-	Mem197[es_186:bx_185 + 0x01:byte] = (byte) (int32) SLICE(SEQ(ax_142, ax_134 + wArg02) + _, word16, 16)
+	Mem0[Mem0[ptrLoc06 + 0x00:byte] + 0x00:byte] = (byte) fn0800_8C8A((word16) dx_ax_177, 0x08, SLICE(dx_ax_177, word16, 16), bp, out dx_181)
```
Comment (uxmal) [inline subjects/regressions/reko-90/PP.c]: This is an improvement...

File: subjects/regressions/reko-90/PP.c [hunk 2]
```diff
@@ -4986,17 +4972,40 @@ byte fn0800_35A3(selector ds, word16 wArg06, word16 wArg08)
...
-		word16 bx_56 = (word16) ptrArg02;
-		selector es_57 = SLICE(ptrArg02, selector, 16);
-		Mem60[es_57:bx_56 + 0x00:word16] = wLoc18;
-		Mem63[es_57:bx_56 + 0x02:word16] = wLoc16;
-		return ds;
-	}
-	else
-		return ds;
+fn0800_363D_entry:
+	def fp
```
Comment (uxmal) [inline subjects/regressions/reko-90/PP.c]: ...but this is a regression

File: subjects/regressions/reko-90/PP.c [hunk 3]
```diff
@@ -10021,26 +10018,37 @@ byte fn0800_8B5A(selector ds, word16 wArg02, ptr16 & siOut, ptr16 & dsOut)
...
-	byte dh_7;
-	word16 cx_8;
-	word16 dx_11 = DPB(dx, msdos_get_system_date(out dh_7, out cx_8), 0);
-	word16 bx_14 = (word16) ptrArg02;
-	selector es_15 = SLICE(ptrArg02, selector, 16);
-	Mem16[es_15:bx_14 + 0x00:word16] = cx_8;
-	Mem17[es_15:bx_14 + 0x02:word16] = dx_11;
-	return;
+fn0800_8B95_entry:
+	def dx
```
Comment (uxmal) [inline subjects/regressions/reko-90/PP.c]: This also is a regression; there should be no `def`s in the generated C.

File: subjects/regressions/reko-90/PP.c [hunk 4]
```diff
@@ -10021,26 +10018,37 @@ byte fn0800_8B5A(selector ds, word16 wArg02, ptr16 & siOut, ptr16 & dsOut)
...
-	word16 bx_15 = (word16) ptrArg02;
-	selector es_16 = SLICE(ptrArg02, selector, 16);
-	byte ch_7;
-	byte dh_8;
-	byte dl_9;
-	Mem17[es_16:bx_15 + 0x00:word16] = DPB(cx, msdos_get_system_time(out ch_7, out dh_8, out dl_9), 0);
-	Mem19[es_16:bx_15 + 0x02:word16] = dx;
-	return;
+fn0800_8BA8_entry:
+	def cx
```
Comment (uxmal) [inline subjects/regressions/reko-90/PP.c]: Another regression

File: subjects/regressions/reko-90/PP.c [hunk 5]
```diff
@@ -10452,22 +10460,58 @@ byte fn0800_8F7F(byte al, selector ds, segptr32 ptrArg02)
...
-		word16 bx_80 = (word16) ptrArg02;
-		selector es_81 = SLICE(ptrArg02, selector, 16);
-		Mem84[es_81:bx_80 + 0x02:word16] = dx_48;
-		Mem85[es_81:bx_80 + 0x00:word16] = ax_51;
-	}
-	byte dl_69;
-	*dlOut = (byte) dx_48;
-	return ax_51;
+fn0800_9764_entry:
+	def fp
```
Comment (uxmal) [inline subjects/regressions/reko-90/PP.c]: Another regression. I have a feeling that all these regressions are related and will likely be fixed by a single change.

File: subjects/regressions/reko-90/PP.c [hunk 6]
```diff
@@ -11243,47 +11287,40 @@ byte fn0800_A614(word16 bp, selector ds, segptr32 ptrArg02, word16 wArg04, ptr16
...
 					return (byte) si_152;
 			}
-			word16 bx_165 = (word16) ptrArg02;
-			selector es_166 = SLICE(ptrArg02, selector, 16);
-			if ((Mem0[es_166:bx_165 + 0x02:word16] & 0x04) != 0x00)
+			if ((Mem0[ptrArg02 + 0x02:word16] & 0x04) != 0x00)
 			{
 				byte dl_177;
-				*dlOut = fn0800_9E75(ds, Mem0[es_166:bx_165 + 0x0A:word16], out ds);
+				*dlOut = fn0800_9E75(ds, Mem0[ptrArg02 + 0x0A:word16], out ds);
```
Comment (uxmal) [inline subjects/regressions/reko-90/PP.c]: ...and I do like this simplification.

**Comment by ptomin** [comment]
John, the result of investigation is 366521f

**Comment by ptomin** [comment]
The reason of regressions is `SliceGegmentedPointer` rule. Memory id is not added to ssa identifiers list.
It causes exception at later stages. I have disabled this rule. Lets think what to do with that

**Comment by uxmal** [comment]
Pavel, could you do me a great favor and write a unit test that shows that memory ID is not added? If you could create a separate PR, I can investigate and fix it. Then we can pull this PR.

**Comment by ptomin** [comment]
OK, I'll write it

**Comment by ptomin** [comment]
John, see #421

**Comment by uxmal** [comment]
Thanks for this Pavel. The PP.EXE regression tests has been improved greatly because of this PR.

**Pull request merged**